nemo/collections/tts/models/tacotron2.py

```diff
@@ -0,0 +1,282 @@
+from nemo.utils import logging
+from nemo.utils.decorators import experimental
+
+
+@dataclass
+class PreprocessorParams:
+    pad_value: float = MISSING
+
+
+@dataclass
+class Preprocessor:
+    cls: str = MISSING
+    params: PreprocessorParams = PreprocessorParams()
+
+
+@dataclass
+class Tacotron2Config:
+    preprocessor: Preprocessor = Preprocessor()
+    encoder: Dict = MISSING
+    decoder: Dict = MISSING
+    postnet: Dict = MISSING
+    labels: List = MISSING
+    train_ds: Optional[Dict] = None
+    validation_ds: Optional[Dict] = None
+
+
+class Tacotron2Loss(Loss):
+    @property
+    def input_types(self):
+        return {
+            "mel_out": NeuralType(('B', 'T', 'D'), MelSpectrogramType()),
+            "mel_out_postnet": NeuralType(('B', 'T', 'D'), MelSpectrogramType()),
+            "gate_out": NeuralType(('B', 'T'), LogitsType()),
+            "mel_target": NeuralType(('B', 'T', 'D'), MelSpectrogramType()),
+            "gate_target": NeuralType(('B', 'T'), LogitsType()),
+            "target_len": NeuralType(('B'), LengthsType()),
+            "pad_value": NeuralType(),
+        }
+
+    @property
+    def output_types(self):
+        return {
+            "loss": NeuralType(elements_type=LossType()),
+        }
+
+    @typecheck()
+    def forward(self, *, mel_out, mel_out_postnet, gate_out, mel_target, gate_target, target_len, pad_value):
+        mel_target.requires_grad = False
+        gate_target.requires_grad = False
+        gate_target = gate_target.view(-1, 1)
+
+        max_len = mel_target.shape[2]
+
+        if max_len < mel_out.shape[2]:
+            # Predicted len is larger than reference
+            # Need to slice
+            mel_out = mel_out.narrow(2, 0, max_len)
+            mel_out_postnet = mel_out_postnet.narrow(2, 0, max_len)
+            gate_out = gate_out.narrow(1, 0, max_len).contiguous()
+        elif max_len > mel_out.shape[2]:
+            # Need to do padding
+            pad_amount = max_len - mel_out.shape[2]
+            mel_out = pad(mel_out, (0, pad_amount), value=pad_value)
+            mel_out_postnet = pad(mel_out_postnet, (0, pad_amount), value=pad_value)
+            gate_out = pad(gate_out, (0, pad_amount), value=1e3)
+            max_len = mel_out.shape[2]
+
+        mask = ~get_mask_from_lengths(target_len, max_len=max_len)
+        mask = mask.expand(mel_target.shape[1], mask.size(0), mask.size(1))
+        mask = mask.permute(1, 0, 2)
+        mel_out.data.masked_fill_(mask, pad_value)
+        mel_out_postnet.data.masked_fill_(mask, pad_value)
+        gate_out.data.masked_fill_(mask[:, 0, :], 1e3)
+
+        gate_out = gate_out.view(-1, 1)
+        mel_loss = nn.MSELoss()(mel_out, mel_target) + nn.MSELoss()(mel_out_postnet, mel_target)
+        gate_loss = nn.BCEWithLogitsLoss()(gate_out, gate_target)
+        return mel_loss + gate_loss
+
+
+@experimental  # TODO: Need to implement abstract methods: list_available_models, from_pretrained, export but how?
```

Comment:
you'll only need to implement "list_available_model()" once you have cloud locations

Reply:
Now I get `TypeError: Can't instantiate abstract class Tacotron2Model with abstract methods export`

Reply:
And `hydra.errors.HydraException: Error calling 'nemo.collections.tts.modules.tacotron2.Encoder' : Can't instantiate abstract class Encoder with abstract methods restore_from, save_to`

Reply:
"Can't instantiate abstract class Tacotron2Model with abstract method" this should not happen if you merge/rebase latest candidate into your branch

Reply:
actually, I might be wrong about this one